monitor/services/src/main/java/org/datacleaner/monitor/server/SchedulingServiceImpl.java

```diff
@@ -131,16 +137,80 @@ private HotFolderAlterationListener(final JobIdentifier job, final ScheduleDefin
+                        file.getAbsolutePath(), job.getName());
+                return;
+            }
+
+            final String dataStoreName = getDataStoreName();
 
-            triggerJobExecution();
+            if (dataStoreName != null) {
+                try {
+                    waitForCompleteFile(file);
+                } catch (final IncompleteFileException e) {
+                    logger.error("Hot folder job execution failed because the file '{}' is incomplete of unavailable. ",
+                            file.getAbsolutePath());
+                    return;
+                }
+
+                final Map<String, String> propertiesMap = new HashMap<>();
+                propertiesMap.put("datastoreCatalog." + getDataStoreName() + ".filename", file.getAbsolutePath());
+                triggerExecution(tenant, job, propertiesMap, TriggerType.HOTFOLDER);
+            } else {
+                triggerJobExecution();
+            }
+        }
+
+        /**
+         * File event can be triggered before the complete file content is written (copying in progress).
+         * This method tries to check that possibility and wait for the complete file to be ready.
+         * Unfortunately it is not file/operating system independent.
+         */
+        private void waitForCompleteFile(final File file) throws IncompleteFileException {
+            final WaitForCompleteFileStrategy waitStrategy;
+            final String osName = System.getProperty("os.name").toLowerCase();
+
+            if (osName.contains("windows")) {
+                logger.info("Using WindowsWaitForCompleteFileStrategy. ");
+                waitStrategy = new WindowsWaitForCompleteFileStrategy();
+            } else if (osName.contains("linux")) {
+                logger.info("Using LinuxWaitForCompleteFileStrategy. ");
+                waitStrategy = new LinuxWaitForCompleteFileStrategy();
+            } else {
+                logger.info("Using GeneralWaitForCompleteFileStrategy. ");
+                waitStrategy = new GeneralWaitForCompleteFileStrategy();
+            }
+
+            waitStrategy.waitForComplete(file);
+        }
+
+        private String getDataStoreName() {
+            try {
+                final TenantContext tenantContext = _tenantContextFactory.getContext(tenant);
+                final JobContext jobContext = tenantContext.getJob(job);
+                final String jobFileName = jobContext.getJobFile().getName();
+                final RepositoryFile jobRepositoryFile = tenantContext.getTenantRootFolder().getFolder("jobs")
+                        .getFile(jobFileName);
+                final JaxbJobReader reader = new JaxbJobReader(tenantContext.getConfiguration());
+                final File jobFile = new File(jobRepositoryFile.toResource().getQualifiedPath());
+
+                return reader.readMetadata(jobFile).getDatastoreName();
```

Comment:
Wow, that's a lot of code for getting the datastore name. Don't we have some sort of utility function that does this?